### PR TITLE
Document all columns in the database

### PR DESCRIPTION
**Pull request by nygrenh**

### CODE REVIEW

**Review by Heliozoa — CHANGES_REQUESTED** [review]

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN regrading_submissions.submission_id IS 'The submission that is regraded in the regrading operation specified in regrading_id.';
+COMMENT ON COLUMN regrading_submissions.grading_before_regrading IS 'The grading object that was set to submission.grading_id before the regrading operation started i.e. the grading where the submission got its points before the regrading operation started.';
+COMMENT ON column regrading_submissions.grading_after_regrading IS 'The grading object that was created in this grading operation.';
+-- users
+COMMENT ON COLUMN users.id IS 'A unique, stable identifier for the record.';
+COMMENT ON COLUMN users.created_at IS 'Timestamp when the record was created';
+COMMENT ON COLUMN users.updated_at IS 'Timestamp when the record was last updated. The field is updated automatically by the set_timestamp trigger.';
+COMMENT ON COLUMN users.deleted_at IS 'Timestamp when the record was deleted. If null, the record is not deleted.';
+COMMENT ON COLUMN users.upstream_id IS 'User''s id in upstream system (currently TMC). Only used to interact with the upstream authentication system. Internally the preferred identifier is user.id.';
```
Comment (nygrenh) [inline services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:113]:
```suggestion
COMMENT ON COLUMN users.upstream_id IS 'User''s id in upstream system (currently TMC). Only used to interact with the upstream authentication system. Internally the preferred identifier is user.id.';
COMMENT ON COLUMN users.email IS 'Email address where the student can be reached. Can change, don''t use this as an identifier.';
```

**Review by george-misan — CHANGES_REQUESTED** [review]

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN exercise_tasks.updated_at IS 'Timestamp when the record was last updated. The field is updated automatically by the set_timestamp trigger.';
+COMMENT ON COLUMN exercise_tasks.deleted_at IS 'Timestamp when the record was deleted. If null, the record is not deleted.';
+COMMENT ON COLUMN exercise_tasks.exercise_id IS 'Exercise where this task is an option.';
+COMMENT ON COLUMN exercise_tasks.exercise_type IS 'Identifier for the exercise service type used for this task. The same identifier needs to exist in exercise_services.slug for the exercise task to work.';
+COMMENT ON COLUMN exercise_tasks.assignment IS 'Exercise assignment (e.g. "In this exercise solve prove the statement P=NP.") in an abstract form. Uses exactly the same format as pages.content.';
+COMMENT ON COLUMN exercise_tasks.private_spec IS 'Fully defines an exercise task: tells how the exercise is structured and how it should be checked for correctness. All other specs can be derived from this one. The system will never share this spec to students because it contains all correctness checking specifications, including the ones that are meant for server-only checks.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:59]: Defines an exercise task fully

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN chapters.name IS 'A human readable name of the chapter. Shown in the chapter lists.';
+COMMENT ON COLUMN chapters.course_id IS 'Which course this chapter belongs to.';
+COMMENT ON COLUMN chapters.front_page_id IS 'Page which is the front page of the chapter i.e. the page where the students end up when they select a chapter in the course material.';
+COMMENT ON COLUMN chapters.opens_at IS 'If set, the chapter is closed to students before the timestamp and opens automatically at the given time. Students are informed of the opening time. If null, the chapter is always open.';
+-- gradings
+COMMENT ON COLUMN gradings.id IS 'A unique, stable identifier for the record.';
+COMMENT ON COLUMN gradings.created_at IS 'Timestamp when the record was created';
+COMMENT ON COLUMN gradings.updated_at IS 'Timestamp when the record was last updated. The field is updated automatically by the set_timestamp trigger';
+COMMENT ON COLUMN gradings.deleted_at IS 'Timestamp when the record was deleted. If null, the record is not deleted.';
+COMMENT ON COLUMN gradings.submission_id IS 'The submission that this grading grades.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:76]: The submission being graded

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN gradings.course_id IS 'The course of this grading. Exists to make certain SQL queries easier. The same information can be obtained though grading.submission.course_instance.course_id.';
+COMMENT ON COLUMN gradings.exercise_id IS 'Exercise this grading is related to. Here to make certain SQL queries easier. The same information can be obtained though gradings.exercise_task.exercise_id.';
+COMMENT ON COLUMN gradings.grading_priority IS 'If there is queue with the gradings, processing priority where gradings with lower value will be processed first.';
+COMMENT ON COLUMN gradings.grading_progress IS 'How the grading progress is progressing. See https://rage.github.io/secret-project-331/headless_lms_actix/models/exercises/enum.GradingProgress.html.';
+COMMENT ON COLUMN gradings.score_given IS 'How many points were given from in this grading. This has been scaled to the max point that are specified in the exercise editor.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:81]: How many points were given in this grading. This has been scaled to the max points of the exercise.

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+-- column will be renamed in the next pr
+COMMENT ON COLUMN gradings.unscaled_max_points IS 'Score maximum we got from exercise service when doing to grading. This has not been adjusted to the max score that was specified in the exercise editor.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:83]: Score maximum we got from the exercise service when doing the grading. This has not been adjusted to the max score is specified in the exercise.

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+-- column will be renamed in the next pr
+COMMENT ON COLUMN gradings.unscaled_score_maximum IS 'How many points exercise service wanted to give from this exercise. This has not been adjusted to the max score that was specified in the exercise editor.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:85]: How many points the exercise service wanted to give from this exercise. This has not been adjusted to the max score that was specified in the exercise.

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN regradings.regrading_started_at IS 'When we started to post the grading tasks in this regrading to exercise services';
+COMMENT ON COLUMN regradings.regrading_completed_at IS 'When we got all results from all the grading tasks from exercise services.';
+COMMENT ON COLUMN regradings.total_grading_progress IS 'What is the overall status of the regrading operation. See https://rage.github.io/secret-project-331/headless_lms_actix/models/exercises/enum.GradingProgress.html.';
+-- regrading_submissions
+COMMENT ON COLUMN regrading_submissions.id IS 'A unique, stable identifier for the record.';
+COMMENT ON COLUMN regrading_submissions.created_at IS 'Timestamp when the record was created.';
+COMMENT ON COLUMN regrading_submissions.updated_at IS 'Timestamp when the record was last updated. The field is updated automatically by the set_timestamp trigger.';
+COMMENT ON COLUMN regrading_submissions.deleted_at IS 'Timestamp when the record was deleted. If null, the record is not deleted.';
+COMMENT ON COLUMN regrading_submissions.regrading_id IS 'The regrading operation where the submission specified in submission_id belongs to.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:104]: The regrading operation to which the submission is attached to.

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:113]: User''s id in upstream system (currently TMC). It is used to interact with the upstream authentication system. Internally the preferred identifier is user.id.'

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN user_exercise_states.grading_progress IS 'Tells how grading of this exercise is progressing towards completion. See https://rage.github.io/secret-project-331/headless_lms_actix/models/exercises/enum.GradingProgress.html. Takes into account more holistic picture of the grading progress than gradings.grading_progress. For example, this takes peer reviews into account and other extra requirements into account.';
+COMMENT ON COLUMN user_exercise_states.activity_progress IS 'Tells how close the activity required from the user is from completion. See: https://rage.github.io/secret-project-331/headless_lms_actix/models/exercises/enum.ActivityProgress.html. The activity is not done before the student has completed everything that is required of them (like giving a peer review).';
+COMMENT ON COLUMN user_exercise_states.selected_exercise_task_id IS 'If an exercise has multiple possible exercise tasks and we''re students a random exercise task, we save the given exercise task here so that we can give the same task back if the user reloads the page and so that we can check during submission that the user is not posting an answer to a task that has not been given to them.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:144]: If an exercise has multiple possible exercise tasks and we are giving each student a random exercise task, we save the given exercise task here. This is used for giving the same task back if the user reloads the page. Also, it is also used for checking during submission to ensure the user can not post an answer to a task that has not been given to them.

File: services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+COMMENT ON COLUMN email_deliveries.sent IS 'Whether the email was sent or not. If false, a background daemon will try to send this email message.';
+COMMENT ON COLUMN email_deliveries.user_id IS 'The user to whom the email should be sent to. If the email template contains dynamic portions with user-specific information (like a grade from a course) this user_id will be used to derive the information.';
+-- email_templates
+COMMENT ON COLUMN email_templates.id IS 'A unique, stable identifier for the record.';
+COMMENT ON COLUMN email_templates.created_at IS 'Timestamp when the record was created.';
+COMMENT ON COLUMN email_templates.updated_at IS 'Timestamp when the record was last updated. The field is updated automatically by the set_timestamp trigger.';
+COMMENT ON COLUMN email_templates.deleted_at IS 'Timestamp when the record was deleted. If null, the record is not deleted.';
+COMMENT ON COLUMN email_templates.content IS 'Content of the email in an abstract form. Same structure for the data as in pages.content but the allowed blocks here is limited. The html and plain text versions of the email are generated from this.';
```
[the remark this answers is not in the document]
Comment (nygrenh) [reply services/headless-lms/migrations/20210713050637_add_comments_to_columns.up.sql:183]: Uses the same structure for the data as pages.content but the allowed blocks here is limited

**Pull request merged**